### PR TITLE
feat: drand: do not initialize clients with no configured servers

### PR DESCRIPTION
**Pull request by arajasek**
## Related Issues

Filecoin mainnet used the DrandIncentinet network for beacon values in the early days of the network. The DrandIncentinet network is no longer live.

This is fine, since all we need is the ability to validate the historical entries in the chain. Unfortunately, due to the way the drand module is implemented in Lotus (see #11085), it's hard to break apart the verification logic into its own thing, so we currently try to set up a "full" DrandClient (in a DrandBeacoon) for DrandIncentinet that isn't actually connected to anything.

## Proposed Changes

Only create the DrandClient if the config includes servers to connect to. This resolves issues reported by @snissn around failing to construct DrandClients, but does introduce the somewhat risky possibility that the `client` in a `DrandBeacon` might be nil.

If we don't like that, we can close this PR without merging, but `import-chain` won't work on mainnet until #11085 is resolved.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: chain/beacon/drand/drand.go [hunk 1]
```diff
@@ -50,6 +50,7 @@ func (dp *drandPeer) IsTLS() bool {
 //
 // The root trust for the Drand chain is configured from build.DrandChain.
 type DrandBeacon struct {
+	// Nil for beacons belonging to "retired" networks
```
[the remark this answers is not in the document]
Comment (ZenGround0) [reply chain/beacon/drand/drand.go:53]: Looks like public key is in the config

[the remark this answers is not in the document]
Comment (ZenGround0) [reply chain/beacon/drand/drand.go:53]: So we still expect retired networks to be in the config, we just drop servers for them?

File: chain/beacon/drand/drand.go [hunk 2]
```diff
@@ -91,32 +92,35 @@ func NewDrandBeacon(genesisTs, interval uint64, ps *pubsub.PubSub, config dtypes
...
-	} else {
-		log.Info("drand beacon without pubsub")
-	}
+		if ps != nil {
+			opts = append(opts, gclient.WithPubsub(ps))
+		} else {
+			log.Info("drand beacon without pubsub")
+		}
 
-	client, err := dclient.Wrap(clients, opts...)
```
[the remark this answers is not in the document]
Comment (arajasek) [reply chain/beacon/drand/drand.go:117]: Yes, that is correct.

**Comment by arajasek** [comment]
> I don't fully understand how this resolves the issue @snissn ran into with `--halt-after-import`. Mostly I don't understand how that case fails but ordinary syncing nodes do not fail.

That's a very good question. I chased that down a bit, and comes down to [this error](https://github.com/drand/drand/blob/master/client/client.go#L53), which is thrown if no clients are specified and the `watcher` is nil. When we ordinarily sync, the watcher is NOT nil, because we supply the [WithPubsub option](https://github.com/filecoin-project/lotus/blob/09b566dde1c3b44f2567679a2049ae8210f2ba1f/chain/beacon/drand/drand.go#L112), which in turn [sets the watcher](https://github.com/drand/drand/blob/master/lp2p/client/client.go#L40). I don't fully understand how correct this is.

> Would it be more correct to upstream changes to drand to separate out verification capabilities from beacon fetching?

I personally don't think that's more correct, though I have somewhat limited understanding here. "Verification capabilities" isn't really a thing the drand _client_ does. It's in the same codebase, but it's effectively a static function: see the method [here](https://github.com/drand/drand/blob/v1.4.9/chain/verify.go#L37-L44), and how Lotus calls it [here](https://github.com/filecoin-project/lotus/blob/09b566dde1c3b44f2567679a2049ae8210f2ba1f/chain/beacon/drand/drand.go#L202).


> > but does introduce the somewhat risky possibility that the client in a DrandBeacon might be nil.
> 
> Main risk here is all SPs upgrade to code and can't make progress because their client is set to nil right? So testplan should probably include mainnet SPs running this code and successfully producing a block.

I'm actually more concerned about some future change, in which this being nil causes a panic in the Lotus node (likely only Lotus nodes connected to miners). I'm confident that won't happen today, it's just not very safe long-term.

**Comment by arajasek** [comment]
Closing as unnecessary.

**Pull request closed without merging**